api/nodes/permissions.py

```diff
@@ -15,6 +17,34 @@ def has_object_permission(self, request, view, obj):
             return obj.can_edit(auth)
 
 
+class AdminOrPublic(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be a Node or User, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        else:
+            return node.has_permission(auth.user, osf_permissions.ADMIN)
+
+
+class ContributorDetailPermissions(permissions.BasePermission):
+    """Permissions for contributor detail page."""
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be User or Node, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
+        user = User.load(request.parser_context['kwargs']['user_id'])
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        elif request.method == 'DELETE':
+            return node.has_permission(auth.user, osf_permissions.ADMIN) or auth.user == user
```

Comment:
Thanks for clarifying the key requirement (ability to self-remove). So long as the permissions check is used in combination with the `get_object` implementation on the view we're ok, though I wouldn't go reusing this permissions class too widely. 

Based on your question, I think there are two unit tests that would be helpful:
1. Test self-removal of a non-admin user from a project (as you describe)
2. The case where an admin can remove themselves, so long as the project has a second admin (we already [have](https://github.com/CenterForOpenScience/osf.io/pull/4235/files#diff-14619b95d1761b6ed098d90acc33b79eR1373) "admin can't remove self when only admin")